File: packages/ts/tests/integration/binding/declared-table-name.test.ts
// Binding-tier tests (real core, real fs) for the declared table `name`.
//
// A table's name is declared, never derived from `ddl`. A `[[table]]` entry
// without `name` fails to load, and a `name` the entry's `ddl` never creates
// fails at load time -- checked against SQLite's own catalog, before any
// ingestion.

import { mkdir, mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { DirSQL } from "dirsql";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

// Emits the file's root-relative `path`. `{path}` is the absolute path,
// `{root}` the index root.
const pathHook = `on-file = '''sh -c 'rel=\${1#"$2"/}; printf "[{\\"path\\":\\"%s\\"}]" "$rel"' sh {path} {root}'''`;

async function seedFile(path: string, content: string): Promise<void> {
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, content);
}

describe("declared [[table]] name", () => {
  let dir: string;
  let configPath: string;

  beforeEach(async () => {
    dir = await mkdtemp(join(tmpdir(), "dirsql-declared-name-"));
    configPath = join(dir, ".dirsql.toml");
    await seedFile(join(dir, "data", "a.csv"), "anything");
  });

  afterEach(async () => {
    await rm(dir, { recursive: true, force: true });
  });

  it("registers a config table under its declared name", async () => {
    await seedFile(
      configPath,
      `
[[table]]
name = "notes"
ddl = "CREATE TABLE notes (path TEXT)"
glob = "data/*.csv"
${pathHook}
`,
    );

    const db = new DirSQL({ root: dir, config: configPath });
    const rows = await db.query("SELECT path FROM notes");
    expect(rows.map((r) => r.path)).toEqual(["data/a.csv"]);
  });

  it("rejects a [[table]] entry with no name", async () => {
    await seedFile(
      configPath,
      `
[[table]]
ddl = "CREATE TABLE notes (path TEXT)"
glob = "data/*.csv"
${pathHook}
`,
    );

    const db = new DirSQL({ root: dir, config: configPath });
    await expect(db.ready).rejects.toThrow(/name/);
  });

  it("rejects a declared name the ddl never creates", async () => {
    await seedFile(
      configPath,
      `
[[table]]
name = "messages"
ddl = "CREATE TABLE notes (path TEXT)"
glob = "data/*.csv"
${pathHook}
`,
    );

    const db = new DirSQL({ root: dir, config: configPath });
    await expect(db.ready).rejects.toThrow(/table 'messages'/);
  });
});

File: packages/python/tests/integration/binding/declared_table_name_test.py
"""Binding-tier tests (real core, real fs) for the declared table `name`.

A table's name is *declared*, never derived from ``ddl``. ``Table`` takes it
as a required keyword argument, a ``[[table]]`` entry without ``name`` fails
to load, and a ``name`` the entry's ``ddl`` does not actually create fails at
load time -- checked against SQLite's own catalog, before any ingestion.
"""

import os
import tempfile

import pytest

from dirsql import DirSQL, Table

# An `on-file` hook emitting the file's root-relative `path`.
_HOOK_PATH = r"""on-file = '''sh -c 'rel=${1#"$2"/}; printf "[{\"path\":\"%s\"}]" "$rel"' sh {path} {root}'''"""


@pytest.fixture
def config_dir():
    with tempfile.TemporaryDirectory() as d:
        yield d


def _fixture(config_dir, config):
    os.makedirs(os.path.join(config_dir, "data"), exist_ok=True)
    with open(os.path.join(config_dir, "data", "a.csv"), "w") as f:
        f.write("anything")
    path = os.path.join(config_dir, ".dirsql.toml")
    with open(path, "w") as f:
        f.write(config)
    return path


def describe_declared_table_name():
    def it_exposes_the_declared_name_on_the_table():
        table = Table(
            name="notes",
            ddl='CREATE TABLE "notes" (path TEXT)',
            glob="data/*.csv",
            on_file=lambda path: [],
        )
        assert table.name == "notes"

    @pytest.mark.asyncio
    async def it_registers_a_config_table_under_its_declared_name(config_dir):
        cfg = _fixture(
            config_dir,
            """\
[[table]]
name = "notes"
ddl = "CREATE TABLE notes (path TEXT)"
glob = "data/*.csv"
"""
            + _HOOK_PATH
            + "\n",
        )

        db = DirSQL(root=config_dir, config=cfg)
        await db.ready()
        results = await db.query("SELECT path FROM notes")
        assert [r["path"] for r in results] == ["data/a.csv"]

    @pytest.mark.asyncio
    async def it_errors_when_a_table_entry_has_no_name(config_dir):
        cfg = _fixture(
            config_dir,
            """\
[[table]]
ddl = "CREATE TABLE notes (path TEXT)"
glob = "data/*.csv"
"""
            + _HOOK_PATH
            + "\n",
        )

        db = DirSQL(root=config_dir, config=cfg)
        with pytest.raises(Exception, match="name"):
            await db.ready()

    @pytest.mark.asyncio
    async def it_errors_when_the_ddl_never_creates_the_declared_name(config_dir):
        cfg = _fixture(
            config_dir,
            """\
[[table]]
name = "messages"
ddl = "CREATE TABLE notes (path TEXT)"
glob = "data/*.csv"
"""
            + _HOOK_PATH
            + "\n",
        )

        db = DirSQL(root=config_dir, config=cfg)
        with pytest.raises(Exception, match="table 'messages'"):
            await db.ready()
